plugins/main/tap/tap.go

```diff
@@ -0,0 +1,472 @@
+// Copyright 2015 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"runtime"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/opencontainers/selinux/go-selinux"
+	"golang.org/x/sys/unix"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	current "github.com/containernetworking/cni/pkg/types/100"
+	"github.com/containernetworking/cni/pkg/version"
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/plugins/pkg/ip"
+	"github.com/containernetworking/plugins/pkg/ipam"
+	"github.com/containernetworking/plugins/pkg/ns"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+	"github.com/containernetworking/plugins/pkg/utils/sysctl"
+)
+
+const AddLink = "AddLink"
+
+type NetConf struct {
+	types.NetConf
+	MultiQueue bool   `json:"multiQueue"`
+	MTU        int    `json:"mtu"`
+	Mac        string `json:"mac,omitempty"`
+
+	RuntimeConfig struct {
+		Mac string `json:"mac,omitempty"`
+	} `json:"runtimeConfig,omitempty"`
+}
+
+// MacEnvArgs represents CNI_ARG
+type MacEnvArgs struct {
+	types.CommonArgs
+	MAC types.UnmarshallableString `json:"mac,omitempty"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func loadConf(args *skel.CmdArgs) (*NetConf, string, error) {
+	n := &NetConf{}
+	if err := json.Unmarshal(args.StdinData, n); err != nil {
+		return nil, "", fmt.Errorf("failed to load netconf: %v", err)
+	}
+	if args.Args != "" {
+		e := MacEnvArgs{}
+		err := types.LoadArgs(args.Args, &e)
+		if err != nil {
+			return nil, "", err
+		}
+
+		if e.MAC != "" {
+			n.Mac = string(e.MAC)
+		}
+	}
+
+	if n.RuntimeConfig.Mac != "" {
+		n.Mac = n.RuntimeConfig.Mac
+	}
+
+	return n, n.CNIVersion, nil
+}
+
+func CloseOnExec(fd int) {
+	syscall.CloseOnExec(fd)
+}
+
+func setContainerSeBool() error {
+	output, err := exec.Command("getsebool", "container_use_devices").CombinedOutput()
```

Comment:
Did you check this plugin without selinux? (i.e. linux machine without getsebool command)

Reply:
Sounds a bit like a premature optimization. This is a workaround for a selinux issue (missing label on the plugin binary), so it might not be even needed or be completely different with other security modules.

Reply:
I'm unsure ... the complexity of this PR is - IMO - keeping SELinux happy. 

Shoving those details in the face of the reader (who might not care about it) is something we should avoid imo. Thus I propose to abstract that, and hide it in a separate package.

Reply:
@mmirecki thank you for the clarification. In my opinion, I suppose we don't require to have interface abstraction because this code is not library and it is not reused other code for now.
@maiqueb is there any detailed plan to reuse this code, for example, you will soon create some other CNI plugin which uses selinux?

Reply:
I prefer that the previous one, without module, the reasons are:

1) Currently only tap plugin may require this mechanism (no other consumer)
2) (At least to me,) there is no other security module which may applied to the abstraction

In addition, if we really required such abstraction, then we should put it in more common place (not CNI).

Reply:
> @mmirecki thank you for the clarification. In my opinion, I suppose we don't require to have interface abstraction because this code is not library and it is not reused other code for now. @maiqueb is there any detailed plan to reuse this code, for example, you will soon create some other CNI plugin which uses selinux?

No. I just think code readability is important on any project, not only for libraries / reused code.

The code becomes a lot more readable if anything related to selinux is moved elsewhere - since the reader can choose to focus on those particular details only if desired.

I won't insist further on this.

Reply:
An alternative version with the selinux stuff moved to a separate package.
Please advise which you prefer.